content/issue-2/applicative-parsing.md

```diff
@@ -0,0 +1,716 @@
+# Programming is hard, looking for better abstractions
+
+If you ever wrote a program bigger than a few lines of code you probably know that the bigger
+the program gets - the harder it's to make sure it does exactly what you meant due to how
+different parts can interact. Over time programmers and computer scientists discovered multiple
+ways to manage this complexity. Programming went from physically wiring the instructions to
+writing them as mnemonics with assembly language, to abstractions with variables, labels,
+functions, and so on.
+
+[Functional Programming](https://en.wikipedia.org/wiki/Functional_programming)
+is one of the paradigms for managing complexity with its main idea of writing code as a set of
+functions and using those functions to pass values around instead imperative style of steps
+to perform to mutate some data.
+
+Main goal of this tutorial is to explain the idea of Applicative Functors, why you may want to
+use them, what each method does. Code shown here is going to be similar but slightly different
+from what you would use in a full featured command line argument parsing library - for
+production you would want better error messages, help generation, dynamic completion and so
+on. But the main operations stay the same. Applicative Functors are not limited to arg parsing
+and you can try to follow this tutorial to get some ideas for your own problems.
+
+Imagine you want to create an argument parser that parses your arguments into a struct like this:
+
+```rust
+struct Args {
+    bin: String,
+    jobs: u32,
+}
+```
+
+If you want to write an argument parser purely imperatively, look at the horrors that entails,
+think about how to scale and maintain this:
+
+```rust
+fn parse() -> Args {
+	let bin = None;
+	let jobs = None;
+
+	let args = std::env::args();
+	while let (Some(arg), Some(value)) = (args.next(), args.next()){
+		if arg == "--bin" {
+			bin = Some(value.to_string());
+		}
+		if arg == "--jobs" {
+			jobs = Some(value.parse::<u32>.unwrap()),
+		}
+	};
+
+	Args {
+		name: name.expect("'--bin' not given"),
+		answer: answer.expect("'--jobs' not given"),
+	}
+}
+```
+
+What if you could use functional programming to just describe what you want:
+
+```rust
+fn parse() -> Args {
+    let name = Arg("--bin").parse(|n| n.to_string());
+    let answer = Arg("--jobs").parse(|a| a.parse::<u32>);
+    let Args = name
+      .zip(answer)
+      .map(|(name, answer)| Args { bin, jobs })
+      .run()
+      .unwrap()
+}
+```
+
+Now lets see how we get there.
+
+# Parsing command line options
+
+The main goal for this tutorial is to implement a basic library to parse command line options.
```

Comment:
> The main goal for this tutorial is to

Thats not the main goal, the main goal was to "Main goal of this tutorial is to explain the idea of Applicative Functors"...

Reply:
Actually we can drop this line from here now that the goal is explained above... 
```suggestion
```